jscomp/test/tagged_template_test.js

```diff
@@ -0,0 +1,62 @@
+// Generated by ReScript, PLEASE EDIT WITH CARE
```

Comment:
@tsnobip is this generated code good -- what you expect?

Reply:
the generated code is good except from the inlining part, it actually doesn't get inlined when you don't add an interface file. 

You can't inline it because you don't know how the `sql` function will handle the interpolation, in this case it won't just do a regular string interpolation but wraps the value of the parameters between simple quotes.
```js 
sql`SELECT * FROM users WHERE id = 5` != sql`SELECT * FROM ${table} WHERE id = ${id}`
```

In the original PR, the decision was made to output a tagged template literal in JS too, but it could instead call the tag function and generate:
```js 
sql(["SELECT * FROM ", " WHERE id = ", ""], "users", "5")
```
and of course 
```js 
sql(["SELECT * FROM users WHERE id = 5"]) != sql(["SELECT * FROM ", " WHERE id = ", ""], "users", "5")
```

Reply:
So is `sql` going to go through the string?
If so, then I don't understand the proposal.
I think it would help to write down in the description explicitly what the transformation is.
So starting from
```
sql`SELECT * FROM ${table} WHERE id = ${id}`
```
I would like to know what ReScript code (or pseudo-code) this generates, not the JS. As the JS is the compiler's business and in general we don't have control over what it does or does not inline. (Meaning: we could have control, but that's the extreme case when you give up and change some detailed back-end thing for this feature only).

Reply:
A bit like what's done for JSX4: where the transformation is part of the spec.

Reply:
> So starting from
> 
> ```
> sql`SELECT * FROM ${table} WHERE id = ${id}`
> ```
> 
> I would like to know what ReScript code (or pseudo-code) this generates
This would be the exact equivalent of:
```rescript
sql([ "SELECT * FROM ", " WHERE id = ", ""], [table, id])
```

I guess originally the original PR kept the tagged template in the generated JS in case some other JS tools would do some transformation on the tagged template, but I personally don't see anything against transforming it into the plain function call.

Reply:
Should sql be safe?
Ie take a tuple.
Or better if it's uniform (and take an array)

Reply:
Let's see if I just make this more confusing, but it might be easier to examplify this without using strings as values/return type. 

For example, imagine we want to ensure we can never send queries through our SQL client that aren't via prepared statements. We have an `sqlValue` type that we supply helpers for constructing, that can only represent valid SQL values. We can then use a template literal to do this:
```rescript
type safeSqlOperation

let safeQuery = sql`select * from whatever where age > ${SqlValue.int4(35)}`
// sql returns a `safeSqlOperation`

let results = await sqlClient.query(safeQuery)
```
...where `safeSqlOperation` would be produced by `sql`, and be a prepared statement + values:
```rescript
{
  query: "select * from whatever where age > $1",
  values: [35]
}
```
...which the SQL client then uses to ensure that we're only querying for things we know went through the `sql` tag, and we've therefore made safe.

So, tagged template literals are about providing APIs that feel natural, but where you can control exactly what values go into it and how they're transformed into the final thing.

Reply:
This is also why inlining doesn't make sense in a tagged template literal setting - tagged templates are about feeding each interpolated value through the tag function.

However, for inlining, I'm wondering if this isn't really that big of a deal. Does it happen if you stick another type than `string` into the second `array` argument of the tag function? Like if the type is opaque, a variant, or whatever.

Reply:
> This would be the exact equivalent of:
> 
> ```rescript
> sql([ "SELECT * FROM ", " WHERE id = ", ""], [table, id])
> ```

OK so I guess `sql` is a function tat takes an array of strings and an array of values. What's the problem if that's inlined if one knows the implementation of `sql`?
If it's a function, then it should be OK to inline, right?

Reply:
Asked a friend a few questions and here's the summary:

Subject: Summary of Discussion on SQL Function and Inlining

This is a summary of our ongoing discussion around the implementation and transformation of the SQL function within the ReScript language, specifically its interaction with JavaScript tagged template literals and the implications of function inlining. The purpose is to align everyone's understanding and summarize the main points.

1. The SQL function is designed to handle tagged template literals in JavaScript, which provides a flexible and natural-feeling API for constructing SQL queries. Tagged template literals allow us to process template literals in a custom way, controlling how values are interpolated into the string.

2. @tsnobip proposed that the `sql` function could be a function that takes an array of strings (representing fragments of the SQL query) and an array of values (to be inserted at the respective placeholders). This model would fit well with the prepared statements concept in SQL, providing protection against SQL injection.

3. @cristianoc raised a question about whether this function should be inlined. Inlining is an optimization that replaces a function call with the function's body, potentially improving performance by reducing the overhead of function calls. However, it also requires certain conditions to be safe and effective. A correctly working compiler should ensure semantic transparency - the behavior of the program should not change whether inlining is performed or not.

4. There was a discussion about whether inlining could disrupt the dynamic processing of the `sql` function, as it relies on runtime values due to its use of tagged template literals. However, given a correctly working compiler and a well-understood usage of the `sql` function, inlining should not pose a problem.

5. @zth provided a valuable example of using the `sql` function to construct safe SQL operations, illustrating how tagged template literals could be used in this context. 

6. Other considerations discussed included type safety, readability, and whether the SQL function should take an array or a tuple. These factors will need to be balanced when finalizing the design and implementation of the `sql` function.

In summary, the `sql` function's design must balance several factors, including the flexibility and readability provided by tagged template literals, the need for type safety and prepared statement construction, and the potential implications of compiler optimizations like inlining. The conversation so far has provided valuable insights into these factors, but further discussion is needed to reach a final design decision.

Reply:
Sure, I'd be happy to summarize this discussion. We started with a TypeScript implementation of a tagged template literal:

```typescript
function myTag(strings: TemplateStringsArray, ...expressions: any[]): string {
    let result = '';

    for (let i = 0; i < strings.length; i++) {
        result += strings[i];
        
        if (i < expressions.length) {
            result += expressions[i];
        }
    }

    return result;
}

let x = 5;
let y = 10;

let str = myTag`Hello, ${x + y} world!`;
console.log(str); // Outputs: "Hello, 15 world!"
```

This was translated into a JavaScript version that was functionally equivalent but without type annotations:

```javascript
function myTag(strings, ...expressions) {
    let result = '';

    for (let i = 0; i < strings.length; i++) {
        result += strings[i];
        
        if (i < expressions.length) {
            result += expressions[i];
        }
    }

    return result;
}

let x = 5;
let y = 10;

let str = myTag`Hello, ${x + y} world!`;
console.log(str); // Outputs: "Hello, 15 world!"
```

We then considered a more optimized version using `reduce`:

```javascript
function myTag(strings, ...expressions) {
    return strings.reduce((result, str, i) => result + str + (expressions[i] || ''), '');
}

let x = 5;
let y = 10;

let str = myTag`Hello, ${x + y} world!`;
console.log(str); // Outputs: "Hello, 15 world!"
```

And then inlined that function:

```javascript
let x = 5;
let y = 10;

let str = ((strings, ...expressions) => {
    return strings.reduce((result, str, i) => result + str + (expressions[i] || ''), '');
})(["Hello, ", " world!"], x + y);

console.log(str); // Outputs: "Hello, 15 world!"
```

We then unrolled the `reduce` into a for loop:

```javascript
let x = 5;
let y = 10;

let str = (() => {
    let strings = ["Hello, ", " world!"];
    let expressions = [x + y];
    let result = "";

    for (let i = 0; i < strings.length; i++) {
        result += strings[i];
        if (i < expressions.length) {
            result += expressions[i];
        }
    }

    return result;
})();

console.log(str); // Outputs: "Hello, 15 world!"
```

And then removed the for loop:

```javascript
let x = 5;
let y = 10;

let str = (() => {
    let strings = ["Hello, ", " world!"];
    let expressions = [x + y];

    return strings[0] + expressions[0] + strings[1];
})();

console.log(str); // Outputs: "Hello, 15 world!"
```

Finally, we performed a compiler-like optimization of evaluating expressions and inlining them directly into the string:

```javascript
let str = "Hello, " + (5 + 10) + " world!";
console.log(str); // Outputs: "Hello, 15 world!"
```

This is the most optimized version of the code, but it also loses the flexibility and generality of the original version. The trade-off between performance and flexibility is a common theme in compiler optimization.

Reply:
when it comes to array vs tuple, in theory, using tuples would make things safer, but I don't know any use case of tagged templates where you know beforehand the number of expected arguments, so I'd rather stick to arrays (you can still use unboxed variants or the new untagged variant feature if you want to handle polymorphic arguments)

Reply:
> When I said that the compiler should not inline the result, I meant that it should not inline it as if it was a plain string interpolation

Does the compiler inline as if it was a plain string interpolation?
Just double checking to be sure.

Reply:
> Does the compiler inline as if it was a plain string interpolation? Just double checking to be sure.

yes that's what it does [for now](https://github.com/rescript-lang/rescript-compiler/blob/c1f4f126a9be4e89a06ff24ff0a53c616f3e6392/jscomp/test/tagged_template_test.js#L8):
```js
var query = Tagged_template_libJs.sql`SELECT * FROM users WHERE id = 5`;
```

Reply:
bsc -drawlambda 
Check that your new lambda instruction is generated

Reply:
> bsc -drawlambda Check that your new lambda instruction is generated

how should I use it?
Just running 
```sh
bsc -drawlambda
```
returns nothing.

Reply:
My bad was assuming familiarity with bsc

bsc -drawlambda Foo.res

Reply:
> when it comes to array vs tuple, in theory, using tuples would make things safer, but I don't know any use case of tagged templates where you know beforehand the number of expected arguments, so I'd rather stick to arrays (you can still use unboxed variants or the new untagged variant feature if you want to handle polymorphic arguments)

Instead of talking about tuples, let's look at what the inputs look like:
```
string value string value string value string value string
```
It seems to me that this is going to be expressible with e.g. some binary operators applied over and over again.
If this captures the intended use cases then:
1) we don't need to pass arrays
2) we don't have the problem of array bounds etc
3) no loops as one can generate the application of the binary operators directly
4) it unlocks inlining so that in simple cases, one pays zero cost for user-defined tags.

Reply:
yes this is exactly the input! Your proposal looks appealing, but would it still work when binding to a JS tag function that expects parameters like this?
```js
foo(["string", "string", "string"], value, value)
```

Reply:
> yes this is exactly the input! Your proposal looks appealing, but would it still work when binding to a JS tag function that expects parameters like this?
> 
> ```js
> foo(["string", "string", "string"], value, value)
> ```

"If this captures the intended use cases" means: if the expected use cases can be expressed by defining the binary operator (s) instead of handling an array
So the tag would just define the operator(s), not a function on arrays. The question is whether the expected use cases are naturally expressed as binary operator(s).

Reply:
Maybe there's a cleaner way to express the binary operator. This is one way anyhow.

Reply:
A perhaps cleaner way would be:
```
(string => 'state, 'value => 'state, ('state, 'state) => 'state)
```